src/content/en/fundamentals/engage-and-retain/push-notifications/sending-messages.markdown

```diff
@@ -3,93 +3,343 @@ layout: shared/narrow
+request. 
 
-You must encrypt all messages before sending them. Encryption is enough of a
-specialty, even within software development, that we don't recommend writing
-your own encryption system. Fortunately, there are [a range of push
-libraries](https://github.com/web-push-libs) including our own [Push Encryption
-Library](https://github.com/GoogleChrome/web-push-encryption).
+You can almost think of this interaction as a conversation. 
 
-## Two ways to send a push message {#ways-to-send}
+App server: "Here's a message can you deliver it within 43200 seconds?"
 
-When implementing server code there are two methods of sending messages to
-consider.
+Message server: "I really can't, but I can do it for you in 21600."
+
+If the TTL has a value of 0, the message server should deliver it immediately or 
+as soon as soon as the user agent is available to receive it.
+
+### Crypto-Key header
+
+To validate a message sent by your app server, the message server needs the 
+public key. You send the public key in the Crypto-Key header. It has several 
+parts.  
+  
+<pre>dh=<i>applicationServerKey</i>;p256ecdsa=<i>publicKey</i></pre>  
+  
+For example:  
+  
+<pre>dh=BGEw2wsHgLwzerjvnMTkbKrFRxdmwJ5S\_k7zi7A1coR\_sVjHmGrlvzYpAT1n4NPbioFlQkIrT  
+NL8EH4V3ZZ4vJE,p256ecdsa=BDd3\_hVL9fZi9Ybo2UUzA284WG5FZR30\_95YeZJsiApwXKpNcF1rRPF3foIiBHXRdJI2Qhumhf6\_LFTeZaN</pre>
+  
+The first part (`dh=applicationServerKey`) is the public key, the one we created 
+under Requesting permission and subscribing users. The second part 
+(`p256ecdsa=publicKey`) is the public key, but base64 url encoded. Notice the 
+comma separating the two parts of the crypto-key.  
+  
+**Note:** A bug in Chrome 52 requires that a semicolon be send to the message 
+server in the Crypto-Key header instead of a comma.
+
+### Authorization header
+
+The standards-based way of sending a message is through a system called 
+Voluntary Application Server Identification for Web Push or VAPID. This article 
+is written with the assumption that VAPID is how your app server will send 
+messages to your client. Everything we've covered so far applies to message 
+servers that don't use VAPID (though, those methods may have additional 
+requirements of their own). 
+
+To use VAPID, you need an an Authorization header. It contains four parts:  
+  
+<pre>Bearer &lt;JWTHeader&gt;.&lt;Payload&gt;.&lt;Signature&gt;</pre>
+
+The word Bearer is a literal that must be followed by a space. The remaining 
+portions, which are encrypted and concatenated using a period are a signed JSON 
+web token (JWT). A JWT is a way of sharing a JSON object with a second party in 
+such a way that the sending party can sign it and the receiving party can verify 
+the signature is from the expected sender. Adding the literal Bearer to the 
+front makes it a bearer token.  
+  
+You must encrypt all messages before sending them. Encryption is enough of a 
```

Comment:
I'd move this encryption piece to the section on encrypting payload or rephrase to be more specific to the Authorization header and signing the JWT.